rust/agama-server/src/cert.rs

```diff
@@ -0,0 +1,63 @@
+use openssl::error::ErrorStack;
+use openssl::pkey::{PKey, Private};
+use openssl::rsa::Rsa;
+use openssl::x509::extension::{KeyUsage, SubjectAlternativeName};
+use openssl::x509::{X509NameBuilder, X509};
+
+use openssl::asn1::Asn1Time;
+use openssl::bn::{BigNum, MsbOption};
+use openssl::hash::MessageDigest;
+use openssl::x509::extension::{BasicConstraints, SubjectKeyIdentifier};
+
+// Generate a self-signed SSL certificate
+// see https://github.com/sfackler/rust-openssl/blob/master/openssl/examples/mk_certs.rs
+pub fn create_certificate() -> Result<(X509, PKey<Private>), ErrorStack> {
+    let rsa = Rsa::generate(2048)?;
+    let key = PKey::from_rsa(rsa)?;
+
+    let mut x509_name = X509NameBuilder::new()?;
+    x509_name.append_entry_by_text("O", "Agama")?;
+    x509_name.append_entry_by_text("CN", "localhost")?;
+    let x509_name = x509_name.build();
+
+    let mut builder = X509::builder()?;
+    builder.set_version(2)?;
+    let serial_number = {
+        let mut serial = BigNum::new()?;
+        serial.rand(159, MsbOption::MAYBE_ZERO, false)?;
+        serial.to_asn1_integer()?
+    };
+    builder.set_serial_number(&serial_number)?;
+    builder.set_subject_name(&x509_name)?;
+    builder.set_pubkey(&key)?;
+
+    let not_before = Asn1Time::days_from_now(0)?;
+    builder.set_not_before(&not_before)?;
+    let not_after = Asn1Time::days_from_now(365)?;
```

Comment:
so basically we create certificate that is valid from today for a next year. Question is are we sure we have correct date in agama? if not, then it can be issue with remote browser, that will have different date.

Reply:
Ah, good point. On the other hand the certificate is self-signed so the browser will complain anyway. Wrong date will make the problem just a little bit more serious. :smile:

Reply:
We could use something like
```rust
// Jan 1 1970 00:00 UTC
let not_before = Asn1Time::from_unix(0)?;
// Jan 1 2040 00:00 UTC 
let not_after = Asn1Time::from_str_x509("20400101000000Z")?;
```
This would cover bigger range.

Anyway, I guess this will be reviewed by the security team later so this might change...

Reply:
Uh, wrong date would even make troubles connecting to other remote https resources like SCC. As a solution we should probably run NTP synchronization on the Live ISO...